bin/runtime-common/src/messages_call_ext.rs

```diff
@@ -57,45 +54,53 @@ pub struct UnrewardedRelayerOccupation {
 	pub free_message_slots: MessageNonce,
 }
 
-impl BaseMessagesProofInfo {
-	/// Returns true if `bundled_range` cannot be directly appended to the `best_stored_nonce`
-	/// or if the `bundled_range` is empty (unless we're confirming rewards when unrewarded
-	/// relayers vector is full).
+/// Info about a `ReceiveMessagesProof` call which tries to update a single lane.
+#[derive(PartialEq, RuntimeDebug)]
+pub struct ReceiveMessagesProofInfo {
+	/// Base messages proof info
+	pub base: BaseMessagesProofInfo,
+	/// State of unrewarded relayers vector.
+	pub unrewarded_relayers: UnrewardedRelayerOccupation,
+}
+
+impl ReceiveMessagesProofInfo {
+	/// Returns true if:
+	///
+	/// - either inbound lane is ready to accept bundled messages;
+	///
+	/// - or there are no bundled messages, but the inbound lane is blocked by too many unconfirmed
+	///   messages and/or unrewarded relayers.
 	fn is_obsolete(&self) -> bool {
 		// transactions with zero bundled nonces are not allowed, unless they're message
 		// delivery transactions, which brings reward confirmations required to unblock
 		// the lane
-		if self.bundled_range.is_empty() {
-			let (free_relayer_slots, free_message_slots) = self
-				.unrewarded_relayers
-				.as_ref()
-				.map(|s| (s.free_relayer_slots, s.free_message_slots))
-				.unwrap_or((MessageNonce::MAX, MessageNonce::MAX));
+		if self.base.bundled_range.is_empty() {
 			let empty_transactions_allowed =
 				// we allow empty transactions when we can't accept delivery from new relayers
-				free_relayer_slots == 0 ||
+				self.unrewarded_relayers.free_relayer_slots == 0 ||
 				// or if we can't accept new messages at all
-				free_message_slots == 0;
-			if empty_transactions_allowed {
-				return false
-			}
-
-			return true
+				self.unrewarded_relayers.free_message_slots == 0;
+			
+			return !empty_transactions_allowed;
 		}
 
 		// otherwise we require bundled messages to continue stored range
-		*self.bundled_range.start() != self.best_stored_nonce + 1
+		*self.base.bundled_range.start() != self.base.best_stored_nonce + 1
```

Comment:
Nit: Maybe we could extract `*self.base.bundled_range.start() != self.base.best_stored_nonce + 1` in another method let's say `BaseMessagesProofInfo::is_obsolete()` since it's used both here and for `ReceiveMessagesDeliveryProofInfo`. In order to avoid duplication.